### PR TITLE
#10883 Allow plugins to dismiss posts through the MessageWillBePosted hook

### PR DESCRIPTION
**Pull request by carmo-evan**
#### Summary
Removes a batchActions() that was preventing invalid posts from being removed. Added a listener to a new type of error that'll allow plugins to dismiss posts if they want to.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/10883

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit tests passed
- [x] Ran `make flow` to ensure type checking passed

#### Test Information
This PR was tested on: Chrome 75

### CODE REVIEW

**Comment by carmo-evan** [comment]
Hi @cpoile @crspeller any thoughts on this?

File: src/actions/posts.js [hunk 1]
```diff
@@ -227,22 +227,18 @@ export function createPost(post, files = []) {
...
 
                         // If the failure was because: the root post was deleted or
                         // TownSquareIsReadOnly=true then remove the post
                         if (error.server_error_id === 'api.post.create_post.root_id.app_error' ||
-                            error.server_error_id === 'api.post.create_post.town_square_read_only'
+                            error.server_error_id === 'api.post.create_post.town_square_read_only' ||
+                            error.server_error_id === 'plugin.message_will_be_posted.dismiss_post'
                         ) {
-                            actions.push(removePost(data));
+                            dispatch(removePost(data));
```
Comment (cpoile) [inline src/actions/posts.js]: I don't fully understand why the batched actions prevented the post from being removed -- did we never try sending removePost before this?

Comment (carmo-evan) [reply src/actions/posts.js]: Not with a batch, I don't think. There's in issue open for redux-batch-actions here https://github.com/tshelburne/redux-batched-actions/issues/3

**Comment by carmo-evan** [comment]
> Makes sense. Needs server side changes to work correct?

Yes. It's just waiting on the Redux PR https://github.com/mattermost/mattermost-server/pull/10921

**Pull request merged**